### PR TITLE
New decorator check_dtypes

### PR DESCRIPTION
**Pull request by oualib**
- check_dtypes
 - code optimization
 - and more...

### CODE REVIEW

**Review by tconsigny — APPROVED** [review]

File: verticapy/stats/tools.py [hunk 1]
```diff
@@ -850,8 +810,9 @@ def het_white(vdf: vDataFrame, eps: str, X: list):
 
 
 # ---#
+@check_dtypes
 @save_verticapy_logs
-def jarque_bera(vdf: vDataFrame, column: str, alpha: float = 0.05):
+def jarque_bera(vdf: vDataFrame, column: str, alpha: Union[int, float] = 0.05):
```
Comment (mail4umar) [inline verticapy/stats/tools.py:815]: separate line

File: verticapy/tests/vModel/test_decision_tree_classifier.py [hunk 2]
```diff
@@ -308,10 +308,8 @@ def test_get_params(self, model):
         params = model.get_params()
 
         assert params == {
-            "n_estimators": 1,
```
Comment (mail4umar) [inline verticapy/tests/vModel/test_decision_tree_classifier.py:311]: i am not sure why you are deleting these parameters. please double check. if you have taken these parameters out then its alright.

File: verticapy/tests/vModel/test_dummy_tree_classifier.py [hunk 3]
```diff
@@ -298,17 +298,7 @@ def test_get_attr(self, model):
...
-            "n_estimators": 1,
-            "max_features": "max",
-            "max_leaf_nodes": 1e9,
-            "sample": 1.0,
-            "max_depth": 100,
-            "min_samples_leaf": 1,
-            "min_info_gain": 0,
-            "nbins": 1000,
-        }
+        assert params == {}
```
Comment (mail4umar) [inline verticapy/tests/vModel/test_dummy_tree_classifier.py:301]: not sure why you are deleting this....

File: verticapy/tests/vModel/test_dummy_tree_regressor.py [hunk 4]
```diff
@@ -152,16 +152,7 @@ def test_get_attr(self, model):
...
-            "n_estimators": 1,
-            "max_features": "max",
-            "max_leaf_nodes": 1000000000,
-            "sample": 1.0,
-            "max_depth": 100,
-            "min_samples_leaf": 1,
-            "min_info_gain": 0,
-            "nbins": 1000,
-        }
+        assert model.get_params() == {}
```
Comment (mail4umar) [inline verticapy/tests/vModel/test_dummy_tree_regressor.py:155]: why are there no parameters?

File: verticapy/toolbox.py [hunk 5]
```diff
@@ -76,14 +81,8 @@ def all_comb(X: list):
...
-def arange(start: float, stop: float, step: float):
-    check_types(
-        [
-            ("start", start, [int, float]),
-            ("stop", stop, [int, float]),
-            ("step", step, [int, float]),
-        ]
-    )
+@check_dtypes
+def arange(start: Union[int, float], stop: Union[int, float], step: Union[int, float]):
```
Comment (mail4umar) [inline verticapy/toolbox.py:85]: separate line

File: verticapy/utilities.py [hunk 6]
```diff
@@ -2655,16 +2447,6 @@ def save_to_query_profile(
     ):
         return False
     try:
-        check_types(
```
Comment (mail4umar) [inline verticapy/utilities.py:2658]: decorator missing at top of function line 2409

File: verticapy/vcolumn.py [hunk 7]
```diff
@@ -3618,19 +3544,8 @@ def pie(
  	--------
  	vDataFrame.donut : Draws the donut chart of the vColumn based on an aggregation.
 		"""
-        if isinstance(pie_type, str):
```
Comment (mail4umar) [inline verticapy/vcolumn.py:3621]: shouldnt we keep this to give more freedom to user to enter capitals?

File: verticapy/vdataframe.py [hunk 8]
```diff
@@ -567,7 +561,6 @@ def __add_to_history__(self, message: str):
     what he/she did. This method is to use to add a customized message in the 
     vDataFrame history attribute.
         """
-        check_types([("message", message, [str])])
```
Comment (mail4umar) [inline verticapy/vdataframe.py:570]: missing decorator on top of fucntion line 557

File: verticapy/vdataframe.py [hunk 9]
```diff
@@ -1590,13 +1582,6 @@ def __vDataFrameSQL__(self, table: str, func: str, history: str):
     ---------------------------------------------------------------------------
     This method is to use to build a vDataFrame based on a relation
         """
-        check_types(
```
Comment (mail4umar) [inline verticapy/vdataframe.py:1593]: decorator needed on top of fucntion line 1580

Comment (mail4umar) [reply verticapy/vdataframe.py:1593]: not necessary i guess because this is an internal function

File: verticapy/vdataframe.py [hunk 10]
```diff
@@ -4484,55 +4340,17 @@ def chaid_columns(self, columns: list = [], max_cardinality: int = 16):
...
     ---------------------------------------------------------------------------
-    Returns a copy of the vDataFrame.
+    Returns a deep copy of the vDataFrame.
 
     Returns
     -------
     vDataFrame
         The copy of the vDataFrame.
         """
-        copy_vDataFrame = vDataFrame("", empty=True)
```
Comment (mail4umar) [inline verticapy/vdataframe.py:4494]: please double check if you want to delete this.

File: verticapy/vdataframe.py [hunk 11]
```diff
@@ -6455,13 +6213,16 @@ def first(self, ts: str, offset: str):
             sql_push_ext=self._VERTICAPY_VARIABLES_["sql_push_ext"],
             symbol=self._VERTICAPY_VARIABLES_["symbol"],
         )
-        self.filter("{} <= '{}'".format(ts, first_date))
+        self.filter(f"{ts} <= '{first_date}'")
         return self
 
     # ---#
     @save_verticapy_logs
```
Comment (mail4umar) [inline verticapy/vdataframe.py:6220]: shouldnt there be the check_type decorator?

File: verticapy/vdataframe.py [hunk 12]
```diff
@@ -10046,9 +9594,9 @@ def scatter_matrix(self, columns: list = [], **style_kwds):
     def search(
```
Comment (mail4umar) [inline verticapy/vdataframe.py:9594]: shouldnt there be the check_dtype decorator here>

File: verticapy/learn/model_selection.py [hunk 13]
```diff
@@ -922,7 +897,7 @@ def enet_search_cv(
     An object containing the result. For more information, see
     utilities.tablesample.
     """
-    check_types([("estimator_type", estimator_type, ["logit", "enet", "auto"])])
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/model_selection.py:925]: solved. good one!

File: verticapy/learn/model_selection.py [hunk 14]
```diff
@@ -2972,18 +2902,10 @@ def stepwise(
 
     if isinstance(X, str):
         X = [X]
-    if isinstance(x_order, str):
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/model_selection.py:2975]: we do it for some function when it is not sensitive but for some others, we will need to transform the method again and again. Let's see how I can easily make it happen in a next PR

File: verticapy/learn/naive_bayes.py [hunk 15]
```diff
@@ -85,18 +90,20 @@ class NaiveBayes(MulticlassClassifier):
...
     @save_verticapy_logs
-    def __init__(self, name: str, alpha: float = 1.0, nbtype: str = "auto"):
-        nbtype_vals = ["auto", "bernoulli", "categorical", "multinomial", "gaussian"]
-        check_types(
-            [
-                ("name", name, [str]),
-                ("alpha", alpha, [int, float]),
-                ("nbtype", nbtype, nbtype_vals),
-            ]
+    def __init__(self, name: str, alpha: Union[int, float] = 1.0, nbtype: str = "auto"):
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/naive_bayes.py:95]: black is automatically indenting the code. Maybe I can find some other options.

File: verticapy/learn/neighbors.py [hunk 16]
```diff
@@ -455,8 +442,9 @@ def cutoff_curve(
         )
 
     # ---#
+    @check_dtypes
     def confusion_matrix(
-        self, pos_label: Union[int, float, str] = None, cutoff: float = -1
+        self, pos_label: Union[int, float, str] = None, cutoff: Union[int, float] = -1
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/neighbors.py:447]: same as previous

File: verticapy/learn/neighbors.py [hunk 17]
```diff
@@ -627,15 +614,6 @@ def predict(
         """
         if isinstance(X, str):
             X = [X]
-        check_types(
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/neighbors.py:630]: good one!

File: verticapy/learn/neighbors.py [hunk 18]
```diff
@@ -978,52 +942,49 @@ class KernelDensity(Regressor, Tree):
...
+        }
         if "store" not in kwargs or kwargs["store"]:
             self.verticapy_store = True
         else:
             self.verticapy_store = False
 
     # ---#
-    def fit(self, input_relation: Union[str, vDataFrame], X: list = []):
+    @check_dtypes
+    def fit(self, input_relation: Union[str, vDataFrame], X: Union[str, list] = []):
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/neighbors.py:987]: same as previous

File: verticapy/learn/preprocessing.py [hunk 19]
```diff
@@ -230,16 +227,17 @@ def deploySQL(self):
         return sql.format(", ".join(self.X), self.name)
 
     # ---#
-    def fit(self, input_relation: Union[str, vDataFrame], X: list = []):
+    @check_dtypes
+    def fit(self, input_relation: Union[str, vDataFrame], X: Union[str, list] = []):
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/preprocessing.py:231]: same as previous

File: verticapy/learn/vmodel.py [hunk 20]
```diff
@@ -3281,14 +1907,15 @@ def classification_report(
     report = classification_report
 
     # ---#
-    def confusion_matrix(self, cutoff: float = 0.5):
+    @check_dtypes
+    def confusion_matrix(self, cutoff: Union[int, float] = 0.5):
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/vmodel.py:1911]: same as previous

File: verticapy/learn/vmodel.py [hunk 21]
```diff
@@ -3297,21 +1924,21 @@ def confusion_matrix(self, cutoff: float = 0.5):
 		An object containing the result. For more information, see
 		utilities.tablesample.
 		"""
-        check_types([("cutoff", cutoff, [int, float])])
         return confusion_matrix(self.y, self.deploySQL(cutoff), self.test_relation,)
 
     # ---#
-    def deploySQL(self, cutoff: float = -1, X: list = []):
+    @check_dtypes
+    def deploySQL(self, cutoff: Union[int, float] = -1, X: Union[str, list] = []):
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/vmodel.py:1931]: same as previous

File: verticapy/learn/vmodel.py [hunk 22]
```diff
@@ -3430,15 +2053,6 @@ def predict(
         # Inititalization
         if isinstance(X, str):
             X = [X]
-        check_types(
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/vmodel.py:3433]: good one!

File: verticapy/learn/vmodel.py [hunk 23]
```diff
@@ -3745,19 +2347,20 @@ def classification_report(
     report = classification_report
 
     # ---#
+    @check_dtypes
     def confusion_matrix(
-        self, pos_label: Union[int, float, str] = None, cutoff: float = -1
+        self, pos_label: Union[int, float, str] = None, cutoff: Union[int, float] = -1
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/vmodel.py:2352]: same as previous

File: verticapy/learn/vmodel.py [hunk 24]
```diff
@@ -4769,16 +3343,17 @@ def score(self, method: str = "r2"):
 class Unsupervised(vModel):
 
     # ---#
-    def fit(self, input_relation: Union[str, vDataFrame], X: list = []):
+    @check_dtypes
+    def fit(self, input_relation: Union[str, vDataFrame], X: Union[str, list] = []):
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/vmodel.py:3347]: same as previous

File: verticapy/learn/vmodel.py [hunk 25]
```diff
@@ -5107,21 +3676,25 @@ def deploySQL(
...
     # ---#
+    @check_dtypes
     def deployInverseSQL(
-        self, key_columns: list = [], exclude_columns: list = [], X: list = []
-    ):
+        self,
+        key_columns: Union[str, list] = [],
+        exclude_columns: Union[str, list] = [],
+        X: Union[str, list] = [],
+    ) -> str:
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/vmodel.py:3685]: same as previous

File: verticapy/vdataframe.py (hunk 9 above)
Comment (oualib) [reply verticapy/vdataframe.py:1593]: yes

File: verticapy/vdataframe.py (hunk 10 above)
Comment (oualib) [reply verticapy/vdataframe.py:4494]: yes deep copy is exactly what I want.

File: verticapy/vdataframe.py (hunk 11 above)
Comment (oualib) [reply verticapy/vdataframe.py:6220]: good one

File: verticapy/vdataframe.py (hunk 12 above)
Comment (oualib) [reply verticapy/vdataframe.py:9594]: good one!

**Pull request merged**